### PR TITLE
Stop multiplying by 0 in GH and copying data

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

A few easy optimizations I noticed when profiling where we were doing memset and memcpy

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp [hunk 1]
```diff
@@ -56,7 +56,6 @@ void TimeDerivative<Dim>::apply(
         christoffel_first_kind_3_up,
     const gsl::not_null<Scalar<DataVector>*> lapse,
     const gsl::not_null<tnsr::I<DataVector, Dim>*> shift,
-    const gsl::not_null<tnsr::ii<DataVector, Dim>*> spatial_metric,
```
Comment (knelli2) [inline src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp:59]: Does it make more sense to just make the spatial metric a reference tag then? Do we ever set it to something other than the spatial part of the spacetime metric? If it's a reference tag then you can take it as an argument rather than `not_null`. Then you don't have to make a view every time this is used.

File: src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp [hunk 2]
```diff
@@ -300,17 +300,35 @@ void TimeDerivative<Dim>::apply(
   }
 
   // Equation for dt_pi
-  for (size_t mu = 0; mu < Dim + 1; ++mu) {
+
+  // WARNING: normal_dot_gauge_constraint is rescaled by gamma0!
+  get(*normal_dot_gauge_constraint) *= get(gamma0);
```
Comment (knelli2) [inline src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp]: Can you add a comment here explaining what you're doing? It looks like you're avoiding a multiplication of `normal_spacetime_one_form` for the relevant components because it's (1,0,0,0), but that's not immediately obvious.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.py [hunk 3]
```diff
@@ -14,30 +14,21 @@
```
Comment (knelli2) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.py]: Can you format this file in a previous separate commit? I was shocked that removing a single arg form the c++ code could cause an almost 700 line change in a py file, until I realized it was largely unrelated formatting changes.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp [hunk 4]
```diff
@@ -295,9 +287,9 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.GeneralizedHarmonic.BCBjorhus.Cls",
...
 
-  const DataVector used_for_size(3);
+  // const DataVector used_for_size(3);
 
-  test_with_random_values<1>(used_for_size);
-  test_with_random_values<2>(used_for_size);
-  test_with_random_values<3>(used_for_size);
+  // test_with_random_values<1>(used_for_size);
+  // test_with_random_values<2>(used_for_size);
+  // test_with_random_values<3>(used_for_size);
```
Comment (knelli2) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp]: Remove?

File: tests/Unit/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.py [hunk 5]
```diff
@@ -5,22 +5,20 @@
```
Comment (knelli2) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.py]: Lump this file into the formatting commit as well. Only one function was actually added here.

**Review by knelli2 — CHANGES_REQUESTED** [review]
You can squash these directly.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp (hunk 4 above)
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp]: I should make sure they work :(

File: src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp:59]: I don't think a reference tag works because those point at an entire different object, whereas this points to a subset. I _think_ compute tags can set the `DataVector`s into the spacetime metric, but adding the argument tags in the places is not something I can reasonably prioritize right now. I do agree that it's worth trying to do what you suggested in the long term, though!

**Comment by nilsdeppe** [comment]
Okay, rebased and made the changes. I squashed the re-enabling of the tests into the last commit

File: src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp [hunk 6]
```diff
@@ -301,6 +301,13 @@ void TimeDerivative<Dim>::apply(
 
   // Equation for dt_pi
 
+  // We first compute the n_a contributes but only for a=0 since n_i=0
```
Comment (knelli2) [inline src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp]: `contributions`

File: src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp [hunk 7]
```diff
@@ -301,6 +301,13 @@ void TimeDerivative<Dim>::apply(
...
+  // identically. We also use dt_Pi_{00} as temporary storage to avoid any extra
+  // allocations and multiple normal_dot_gauge_constraint=(n^a C_a) by gamma0
```
Comment (knelli2) [inline src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp]: `multiply`

**Review by knelli2 — COMMENTED** [review]
Just some typos. You can squash it all together

**Review by knelli2 — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
I'll see if I can figure out the python formatting... I think I was still using yapf instead of black. I'll ping you once I got that sorted out :)

**Comment by nilsdeppe** [comment]
Okay, python formatting seems to be sorted out...

**Review by knelli2 — APPROVED** [review]
